SE-0334: Pointer API Usability Improvements

glessard opened the pull request:
Implementation of SE-0334, "Pointer API Usability Improvements. See proposal [here](https://github.com/apple/swift-evolution/blob/main/proposals/0334-pointer-usability-improvements.md).

Addresses rdar://64342031, [SR-11156](https://bugs.swift.org/browse/SR-11156) (rdar://53272880), rdar://22541346 and rdar://88087964.

glessard committed:
[stdlib] `alignedUp(for:)` and siblings for URP and UMRP

glessard committed:
[stdlib] comparisons for heterogeneous pointer types

glessard committed:
[stdlib] `pointer(to property:)` for UP and UMP

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please smoke test macOS platform

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test macOS platform

glessard committed:
[stdlib] restrict pointer arithmetic to UInt

glessard commented:
@swift-ci please smoke test

stephentyrone commented on stdlib/public/core/UnsafeRawPointer.swift:
```diff
@@ -1145,6 +1215,76 @@ extension UnsafeMutableRawPointer: Strideable {
+  public func alignedUp(toMultipleOf alignment: Int) -> Self {
+    let mask = alignment &- 1
+    _debugPrecondition(alignment & mask == 0)
+    let bits = (Int(Builtin.ptrtoint_Word(_rawValue)) &+ mask) & ~mask
```
It's not an issue on most platforms, because addresses very near near the top of the address space are usually reserved and not valid pointers, but I think that formally we want this to be a non-wrapping addition on `UInt` to prevent aligning up from wrapping around to address zero.

glessard committed:
[stdlib] relax stride check

- The stride check in `UnsafePointer.withMemoryRebound` makes less sense when rebinding memory for a single element.
- This skips the stride-matching portion of the `_debugPrecondition` in `withMemoryRebound` when `count == 1`.

glessard committed:
[test] UnsafePointer.withMemoryRebound with capacity 1

glessard commented:
rebased to resolve merge conflict

glessard commented:
@swift-ci please smoke test

glessard committed:
[test] test SE-0334 additions

glessard commented:
@swift-ci please smoke test

glessard committed:
Merge branch 'main' into se-pointer-convenience

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test macOS platform
